### DISCUSSION TITLE
Force native compilation from the Makefile

### DISCUSSION THREAD

**Discussion by DamienCassou** [emacscollective/borg#116] [Ideas]
I'm just discovering Emacs 28. After I cleaned all byte compiled files, I ran `make` (which took some time) and then started Emacs 28. After Emacs had started, I got some messages in the `*Warnings*` buffer which seemed to indicate that Emacs was compiling my packages. I was wondering if `make` could do that so Emacs wouldn't have to do that interactively.

**Comment by tarsius** [comment]
> I ran `make` (which took some time)

If we add native compilation here, then it is going to take much longer.

> After Emacs had started, I got some messages in the `*Warnings*` buffer which seemed to indicate that Emacs was compiling my packages.

This is controlled by `native-comp-verbose`, though the default is to not show any warnings. And the warnings are supposed to go into `*Native-compile-Log*` buffer I think. I am currently on 29, so maybe this is different in 28. That would surprise me though; are you maybe on an *old* 28 build?

> I was wondering if `make` could do that so Emacs wouldn't have to do that interactively.

Looks like it would be as simple as replacing `batch-byte-compile` with `batch-native-compile`, so I should probably at least make this configurable.

**Reply by tarsius** [reply]
Related: https://github.com/emacscollective/auto-compile/issues/25.

**Reply by tarsius** [reply]
Well it is a little more difficult than that, but I basically got it working. The problem is that this revealed some existing strangeness, that I have to think about first before I can make this configurable.

**Reply by tarsius** [reply]
Added new variable `borg-compile-function`, which can be set to `native-compile` to enable upfront native compilation.  This should be done in `~/.emacs.d/etc/borg/config.el` like so:

```elisp
(when (native-comp-available-p)
  (setq borg-compile-function #'native-compile))
```

Still waiting for first native `make clean all` to complete so that I can tell you how much longer it takes...

**Reply by tarsius** [reply]
Something like 40 minutes instead of three.

(I don't know precisely because I had to abort on `yaml.el`. That seems like a pathological case, 10 minutes and counting for a 2800 lines file seems like a bit much.)

**Reply by DamienCassou** [reply]
Thank you very much. I'm not sure this is worth it anymore :-).

**Reply by DamienCassou** [reply]
To me `native-comp-verbose` is set to `0` but I still get warnings (in the `*Warnings*` buffer). The variable `native-comp-async-report-warnings-errors` has value `t` though.

**Reply by tarsius** [reply]
Probably not, but now its possible. :grinning: 

I don't know why your `comp` is more verbose than mine. Turns out I have not customized any of its variables.

**Reply by dakra** [reply]
Just want to say that I really appreciate this new feature.
It's just so much nicer to let everything native-compile over night and then be done with it after an update.

Maybe this could also be a Make target?
I have it in my borg/config.el now but maybe sometimes I only want normal byte compilation and sometimes native.
So `make eln` and `make elc` or something like that could be handy.

And when I native compile from Emacs it spawns 4 Emacs native compilation processes while
borg only does one at a time. Compiling parallel would speed things up (even though not really necessary for me).

Thanks :)

**Reply by tarsius** [reply]
I haven't tried it out at all yet, but the changes in the `build-native` branch might do the trick.

**Reply by dakra** [reply]
Did you forget to push the `build-native` branch to GitHub?

**Reply by tarsius** [reply]
Yes but when you reminded me, then I pushed. Have you had a chance to try it out yet?

**Reply by dakra** [reply]
Hey, I tried it very briefly.
build-native has an escaping error:
```
make build-native
/bin/sh: -c: line 3: syntax error near unexpected token `)'
/bin/sh: -c: line 3: `--eval '(borg-batch-rebuild nil \'native-compile-async)' \'
make: *** [lib/borg/borg.mk:72: build-native] Error 2
```
Changing it to:
```
--eval "(borg-batch-rebuild nil 'native-compile-async)" \
```
seems to work but then I run into a different error.
But this could just have been a problem with the current master branch which
segfaulted for me with or without your change.

Since native-compilation is always a bit time consuming I do it rather irregular, once or twice a month.
I'll post here next time I try it if it's working or not.

Thanks.

**Reply by eki3z** [reply]
@tarsius  same problem as @dakra 

I doubt is there any method to native compile only modifiled file when I update a submodule? When I compile lib/% to elc, it's ok because it doesn't take much timeto compile whole module. But eln just is too time-consumig.

**Reply by tarsius** [reply]
> same problem

I already fixed the quote problem and force pushed. So I assume that by "same problem" you mean the seqfault.

**Reply by dakra** [reply]
> is there any method to native compile only modifiled file when I update a submodule?

If you set `(setq native-comp-deferred-compilation t)` it will native compile the code when loading the elc.

**Reply by Eason0210** [reply]
> ```emacs-lisp
> (when (native-comp-available-p)
>   (setq borg-compile-function #'native-compile))
> ```
I set this in my` etc/borg/config.el`,  and run ` make all` , but I always got failed report ( `Done (Total of 0 files compiled, 2 failed)`).
Do I miss something else ?
```
--- [borg] ---
  Loading c:/Users/Aqua/.emacs.d/etc/borg/config.el...
 Creating c:/Users/Aqua/.emacs.d/lib/borg/borg-autoloads.el...
Compiling c:/Users/Aqua/.emacs.d/lib/borg/borg-elpa.el...
Compiling c:/Users/Aqua/.emacs.d/lib/borg/borg.el...
Done (Total of 0 files compiled, 2 failed)
  Running `makeinfo --no-split borg.texi -o borg.info'...
  Running `makeinfo --no-split borg.texi -o borg.info'...done
  Running `install-info borg.info --dir=dir'...
  Running `install-info borg.info --dir=dir'...done
... more packages ...
```

**Reply by tarsius** [reply]
I can reproduce it with recent 28 and 29 and even the older 28.0.90. So maybe this never worked. Could also be a regression though (in Emacs or Borg).

I don't really have the time or interest to investigate this, but would be interested in the results if you could investigate this yourself.

**Reply by Eason0210** [reply]
> I can reproduce it with recent 28 and 29 and even the older 28.0.90. So maybe this never worked. Could also be a regression though (in Emacs or Borg).
> 
> I don't really have the time or interest to investigate this, but would be interested in the results if you could investigate this yourself.

Hi @tarsius ,

I known why we got the wrong compile message now.

Because `byte-compile-file` will return `t` when compile success , but `native-compile` will return the filename of the
compiled object , not `t`. So the `fail-count` is actually success compile + failed compile.



And when `native-compile` failed, it will not return `nil`, I don't known how to handle this.

The bug is in here: 
https://github.com/emacscollective/borg/blob/master/borg.el#L749-L750

```
native-compile is a native compiled Lisp function in ‘comp.el’.

(native-compile FUNCTION-OR-FILE &optional OUTPUT)

Compile FUNCTION-OR-FILE into native code.
This is the synchronous entry-point for the Emacs Lisp native
compiler.
FUNCTION-OR-FILE is a function symbol, a form, or the filename of
an Emacs Lisp source file.
If OUTPUT is non-nil, use it as the filename for the compiled
object.
If FUNCTION-OR-FILE is a filename, return the filename of the
compiled object.  If FUNCTION-OR-FILE is a function symbol or a
form, return the compiled function.

  Probably introduced at or before Emacs version 28.1.
```
Tested on macOS with [latest commit of Emacs ](https://github.com/emacs-mirror/emacs/commit/2212b42806757957fff6a9646debddecb301241c)

**Reply by tarsius** [reply]
Thanks!

820e5262314036b35a7a5225a763ffc370c9a93a should fix it.

**Reply by Eason0210** [reply]
Great, I tested and It works well now. Thanks!
